### PR TITLE
Change the GKE cluster to a private one (instead of public)

### PR DESCRIPTION
**Pull request by lavjain**
Also change the vpc-name to `vpc-01` (from `default-vpc`)

### CODE REVIEW

File: terraform/stages/gke/main.tf [hunk 1]
```diff
@@ -18,20 +18,45 @@
...
 
+data "terraform_remote_state" "foundation" {
+  backend       = "gcs"
+  config = {
+    bucket = "${var.project_id}-tfstate"
+    prefix = "stage/foundation"
+  }
+}
+
+module "vpc_network" {
```
Comment (jonchenn) [inline terraform/stages/gke/main.tf:34]: VPC network should be created in foundation stage.

File: terraform/modules/vpc_network/main.tf [hunk 2]
```diff
@@ -24,26 +24,21 @@ module "vpc" {
 
   subnets = [
     {
-      subnet_name               = "vpc-01-subnet-01"
+      subnet_name               = var.vpc_subnetwork
       subnet_ip                 = "10.0.0.0/16"
```
Comment (jonchenn) [inline terraform/modules/vpc_network/main.tf]: Should this be a variable too?

**Review by jonchenn — CHANGES_REQUESTED** [review]
Leave a comment on the VPC network creation.

File: terraform/stages/foundation/main.tf [hunk 3]
```diff
@@ -54,6 +54,12 @@ module "service_accounts" {
   project_number = data.google_project.project.number
 }
 
+resource "google_compute_network" "vpc_network" {
```
Comment (jonchenn) [inline terraform/stages/foundation/main.tf]: Why not using the ../../modules/vpc_network?  You'd want to handle multiple scenarios:
- Create a brand new VPC like it was.
- Use an existing shared VPC

File: terraform/stages/gke/main.tf [hunk 4]
```diff
@@ -17,21 +17,58 @@
 
 # project-specific locals
 locals {
-  lh = join("", ["local", "host"])
+  lh             = join("", ["local", "host"])
+  vpc_network    = data.terraform_remote_state.foundation.outputs.vpc_network
+  vpc_network_id = data.terraform_remote_state.foundation.outputs.vpc_network_id
+  api_domain     = data.terraform_remote_state.foundation.outputs.api_domain
```
Comment (jonchenn) [inline terraform/stages/gke/main.tf]: Does api_domain need to be from the `foundation` stage?

File: terraform/stages/gke/variables.tf [hunk 5]
```diff
@@ -58,18 +103,6 @@ variable "storage_multiregion" {
   default = "us"
 }
 
-variable "admin_email" {
-  type = string
-  # TODO: replace with your own email
-  default = "admin@google.com"
-}
-
-variable "api_domain" {
```
Comment (jonchenn) [inline terraform/stages/gke/variables.tf]: ditto: I think the `api_domain` can live in the gke stage, the foundation stage doesn't need the api_domain var.

**Review by jonchenn — CHANGES_REQUESTED** [review]
Added a few comments on the vpc_network and api_domain var.

File: terraform/stages/gke/main.tf (hunk 1 above)
Comment (lavjain) [reply terraform/stages/gke/main.tf:34]: Done

File: terraform/modules/vpc_network/main.tf (hunk 2 above)
Comment (lavjain) [reply terraform/modules/vpc_network/main.tf]: Done

File: terraform/stages/foundation/main.tf (hunk 3 above)
Comment (lavjain) [reply terraform/stages/foundation/main.tf]: vpc_network module does not let you use an existing vpc. I am creating a new VPC for GKE as a workaround

File: terraform/stages/gke/main.tf (hunk 4 above)
Comment (lavjain) [reply terraform/stages/gke/main.tf]: Not anymore

File: setup/setup_all.sh [hunk 6]
```diff
@@ -102,10 +102,15 @@ init_foundation() {
   
   # Enabling GCP services first.
   terraform apply -target=module.project_services -target=module.service_accounts -auto-approve
-  
-  # Initializing Firebase (Only need this for the first time.)
-  # NOTE: the Firebase can only be initialized once (via App Engine).
-  terraform apply -target=module.firebase -var="firebase_init=true" -auto-approve
+
+  # Check if firebase is already initialized
+  KEY_NAME=$(gcloud alpha services api-keys list --filter="displayName='Browser key (auto created by Firebase)'" \
```
Comment (jonchenn) [inline setup/setup_all.sh:108]: I double-checked this method, this is not reliable. The API key will only be created when creating a Firebase app on the Console UI.

File: setup/setup_all.sh [hunk 7]
```diff
@@ -114,6 +119,14 @@ init_foundation() {
 # Build all microservices and deploy to the cluster:
 deploy_microservices_to_gke() {
   cd "$BASE_DIR"/terraform/stages/gke
+
+  # Check for GKE-SA service account and delete if exists
+  GKE_SA=gke-sa@${PROJECT_ID}.iam.gserviceaccount.com
+  GKE_SA_EXISTS=$(gcloud iam service-accounts list | grep -c "$GKE_SA")
+  if [[ "$GKE_SA_EXISTS" == 1 ]]; then
+    gcloud iam service-accounts delete "$GKE_SA" --quite
+  fi
```
Comment (jonchenn) [inline setup/setup_all.sh]: This shouldn't be here. This should be part of the clean up after e2e-test.

File: terraform/modules/gke/main.tf [hunk 8]
```diff
@@ -72,11 +74,51 @@ module "gke_cluster" {
...
+#      role = "general"
+#    }
+#
+#    service_account = google_service_account.kubernetes.email
+#    oauth_scopes = [
+#      "https://www.googleapis.com/auth/cloud-platform"
+#    ]
+#  }
+#}
+
```
Comment (jonchenn) [inline terraform/modules/gke/main.tf]: why comment this out?

File: terraform/modules/ingress_gce/main.tf [hunk 9]
```diff
@@ -34,6 +34,10 @@ locals {
     var.external_ip_address != null ? var.external_ip_address
     : google_compute_global_address.ingress_ip_address[0].name
   )
+  cert_name = (
+    length(kubectl_manifest.managed_certificate) == 0 ? "dummy-cert"
```
Comment (jonchenn) [inline terraform/modules/ingress_gce/main.tf]: maybe call it "default-cert"?

File: terraform/modules/vpc_network/main.tf [hunk 10]
```diff
@@ -17,35 +17,52 @@
...
-        range_name    = "secondary-service-range-01"
-        ip_cidr_range = "10.2.0.0/16"
-      },
+    (var.vpc_subnetwork) = [
+      var.secondary_ranges_pods,
+      var.secondary_ranges_services,
     ]
   }
+
+  firewall_rules = [
```
Comment (jonchenn) [inline terraform/modules/vpc_network/main.tf]: Can you add comment on why we need a firewall rules here?

**Review by jonchenn — CHANGES_REQUESTED** [review]
A few high-level comments:
- The existing gke-sa clean up should not be in the setup scripts. It shall be in the clean up job after e2e test.
- Add more comment on any additional resources added to the gke.

File: terraform/modules/gke/main.tf (hunk 8 above)
Comment (lavjain) [reply terraform/modules/gke/main.tf]: The recommendation is to create a node-pool separately so changing it does not re-create the existing cluster. I will add a TODO comment here

File: terraform/modules/ingress_gce/main.tf (hunk 9 above)
Comment (lavjain) [reply terraform/modules/ingress_gce/main.tf]: Thanks

File: terraform/modules/vpc_network/main.tf (hunk 10 above)
Comment (lavjain) [reply terraform/modules/vpc_network/main.tf]: Will do

File: setup/setup_all.sh (hunk 6 above)
Comment (lavjain) [reply setup/setup_all.sh:108]: Will add a TODO comment to change this with a more reliable method

File: terraform/modules/gke/main.tf [hunk 11]
```diff
@@ -72,11 +74,52 @@ module "gke_cluster" {
   }
 }
 
+# TODO: Define GKE node pool separate from the cluster
```
Comment (jonchenn) [inline terraform/modules/gke/main.tf]: If this is not used anywhere in the gke module now, can you remove it?

File: terraform/modules/gke/main.tf [hunk 12]
```diff
@@ -72,11 +74,52 @@ module "gke_cluster" {
...
+#    ]
+#  }
+#}
+
 resource "time_sleep" "wait_for_gke" {
   depends_on      = [module.gke_cluster]
   create_duration = "120s"
 }
 
+module "cloud-nat" {
```
Comment (jonchenn) [inline terraform/modules/gke/main.tf:83]: Can you add a comment on why we need a Cloud NAT?

File: terraform/modules/shared_vpc/main.tf [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+    }
+  ]
+
+  firewall_rules = [
+    {
+      name                    = "serverless-to-vpc-connector"
+      description             = null
+      direction               = "INGRESS"
+      priority                = null
+      ranges                  = ["107.178.230.64/26", "35.199.224.0/19"]
```
Comment (jonchenn) [inline terraform/modules/shared_vpc/main.tf:51]: The range should be customizable, can you add them into vars?

File: terraform/modules/shared_vpc/main.tf [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+      log_config = {
+        metadata = "EXCLUDE_ALL_METADATA"
+      }
+    },
+    {
+      name                    = "vpc-connector-to-serverless"
+      description             = null
+      direction               = "EGRESS"
+      priority                = null
+      ranges                  = ["107.178.230.64/26", "35.199.224.0/19"]
```
Comment (jonchenn) [inline terraform/modules/shared_vpc/main.tf:80]: ditto, if the ranges are customizable, they shall be vars.

File: terraform/modules/shared_vpc/main.tf [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+#   router                             = google_compute_router.router.name
+#   region                             = google_compute_router.router.region
+#   nat_ip_allocate_option             = "AUTO_ONLY"
+#   source_subnetwork_ip_ranges_to_nat = "ALL_SUBNETWORKS_ALL_IP_RANGES"
+
+#   log_config {
+#     enable = true
+#     filter = "ERRORS_ONLY"
+#   }
+# }
```
Comment (jonchenn) [inline terraform/modules/shared_vpc/main.tf]: Remove codes that are not used.

File: terraform/stages/foundation/main.tf [hunk 16]
```diff
@@ -54,6 +67,16 @@ module "service_accounts" {
   project_number = data.google_project.project.number
 }
 
+module "shared_vpc" {
```
Comment (jonchenn) [inline terraform/stages/foundation/main.tf:70]: where can we choose to use shared_vpc vs. regular vpc_network? Is it through the vars.shared_vpc?

File: terraform/stages/foundation/main.tf [hunk 17]
```diff
@@ -62,10 +85,3 @@ module "firebase" {
   firebase_init    = var.firebase_init
 }
 
-module "vpc_network" {
```
Comment (jonchenn) [inline terraform/stages/foundation/main.tf:65]: Do we not use this module anymore?

File: terraform/stages/foundation/outputs.tf [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+output "vpc_network" {
+  value = module.shared_vpc.network_name
```
Comment (jonchenn) [inline terraform/stages/foundation/outputs.tf:19]: What if we choose to use regular VPC instead of shared_vpc?

**Review by jonchenn — CHANGES_REQUESTED** [review]

File: terraform/stages/foundation/main.tf (hunk 17 above)
Comment (lavjain) [reply terraform/stages/foundation/main.tf:65]: The VPC module is being used in GKE stage

File: terraform/stages/foundation/outputs.tf (hunk 18 above)
Comment (lavjain) [reply terraform/stages/foundation/outputs.tf:19]: The foundation stage will create a common VPC (regular). Use of shared VPC is optional

File: terraform/modules/shared_vpc/main.tf (hunk 13 above)
Comment (frits-v) [reply terraform/modules/shared_vpc/main.tf:51]: These are not variable

File: terraform/modules/gke/main.tf (hunk 11 above)
Comment (lavjain) [reply terraform/modules/gke/main.tf]: Done

**Comment by lavjain** [comment]
Closing this pull request to separate out private cluster and shared vpc parts

**Pull request closed without merging**